### PR TITLE
 rofiles: Add --copyup option

### PR DESCRIPTION
**Pull request by cgwalters**
Sadly https://sourceware.org/bugzilla/show_bug.cgi?id=22089 is I think going to
actually force us to cave here. Even if we got the glibc patch in today, we need
to support the RHEL glibc. See also discussion about fish as part of the general
Fedora tracker.

This is basically needed to unblock rpm-ostree unified core :globe_with_meridians::
projectatomic/rpm-ostree#729

Closes: #1377

### CODE REVIEW

**Comment by cgwalters** [comment]
Depends: https://github.com/ostreedev/ostree/pull/1378

File: src/rofiles-fuse/main.c [hunk 1]
```diff
@@ -286,6 +333,9 @@ callback_truncate (const char *path, off_t size)
 static int
 callback_utimens (const char *path, const struct timespec tv[2])
 {
+  /* This one isn't write-verified, we support changing times
+   * even for hardlinked files.
```
Comment (jlebon) [inline src/rofiles-fuse/main.c]: Can you expand on this? Don't we need to keep bare objects at `OSTREE_TIMESTAMP` to be consistent?

File: src/rofiles-fuse/main.c [hunk 2]
```diff
@@ -324,13 +374,22 @@ do_open (const char *path, mode_t mode, struct fuse_file_info *finfo)
...
         {
           (void) close (fd);
           return r;
         }
 
+      /* In the copyup case, we need to re-open */
+      if (opt_copyup)
+        {
+          (void) close (fd);
+          fd = openat (basefd, path, finfo->flags & ~O_TRUNC, mode);
```
Comment (jlebon) [inline src/rofiles-fuse/main.c]: Minor: we can just stop filtering out `O_TRUNC` in the copy-up case, right? And only pay the `ftruncate` call in the non-copy-up case.

File: src/rofiles-fuse/main.c (hunk 1 above)
Comment (cgwalters) [reply src/rofiles-fuse/main.c]: It's an interesting point...gets a bit into a philosophical question of whether rofiles-fuse is for "anything using hardlinks" or whether it's really specific to ostree.  The "anything using hardlinks" case might be a backup system or non-os-data object store was my thought originally.

For example I could probably have "object semantics" for my `~/Documents` which has a bunch of random PDFs and `.txt` and LibreOffice files etc. that should really never be overwritten in place.  OTOH I've been thinking about storing all of that in git.

So...hmm.  It's not a new issue here and we should think about changing this, but I worry a bit about the backcompat issues.  In practice nothing we use rofiles-fuse for should be touching timestamps for files it didn't create at the time I would guess.

**Comment by cgwalters** [comment]
Consumer: https://github.com/projectatomic/rpm-ostree/pull/1171

File: src/rofiles-fuse/main.c (hunk 1 above)
Comment (dustymabe) [reply src/rofiles-fuse/main.c]: > For example I could probably have "object semantics" for my ~/Documents which has a bunch of random PDFs and .txt and LibreOffice files etc. that should really never be overwritten in place. OTOH I've been thinking about storing all of that in git.

I've been using [git-annex](https://git-annex.branchable.com/) for this for some time now. Good news, it's in fedora.

Comment (cgwalters) [reply src/rofiles-fuse/main.c]: Right, git-annex is likely an excellent example of a possible user of rofiles-fuse semantics, specifically [its direct mode](https://git-annex.branchable.com/direct_mode/).  Though I didn't look closely.

File: src/rofiles-fuse/main.c (hunk 2 above)
Comment (cgwalters) [reply src/rofiles-fuse/main.c]: BTW I wish I'd seen this comment while working on this patchset on the plane :airplane:  - it turned out mishandling `O_TRUNC` was the source of bugs.

File: src/rofiles-fuse/main.c [hunk 3]
```diff
@@ -385,7 +385,12 @@ do_open (const char *path, mode_t mode, struct fuse_file_info *finfo)
       if (opt_copyup)
         {
           (void) close (fd);
-          fd = openat (basefd, path, finfo->flags & ~O_TRUNC, mode);
+          /* Note that unlike the initial open, we will pass through
+           * O_TRUNC.  More ideally in this copyup case we'd avoid copying
```
Comment (jlebon) [inline src/rofiles-fuse/main.c]: Hmm, so then we don't need to `ftruncate` down below again, right? Shouldn't the `if` condition at 399 be changed to

```
if (!opt_copyup && (finfo->flags & O_TRUNC))
```

?

File: src/rofiles-fuse/main.c [hunk 4]
```diff
@@ -385,7 +385,12 @@ do_open (const char *path, mode_t mode, struct fuse_file_info *finfo)
...
+           * the whole file in the first place, but eh.  It's not like we're
+           * high performance anyways.
+           */
+          fd = openat (basefd, path, finfo->flags & ~(O_EXCL|O_CREAT), mode);
```
Comment (jlebon) [inline src/rofiles-fuse/main.c]: Why are we filtering out `O_EXCL` and `O_CREAT` here? This is something that should've been caught by the first `openat`, right? Or is this just to not have a TOCTOU race on the file existing?

Comment (cgwalters) [reply src/rofiles-fuse/main.c]: We're internally reopening the file we just created.  So if the user provided `O_EXCL` we need to filter that out.  We might as well drop `O_CREAT` too.

**Comment by jlebon** [comment]
@rh-atomic-bot r+ eec2ff2

**Pull request closed without merging**